### ISSUE TITLE
[Form] Add Material Design Form Control Helper

### ISSUE THREAD

**Issue by TBSten** [soil-kt/soil#27]
## Overview

Provide a helper for the `material` / `material3` component to the Soil Form.

## Problem

When I was trying to implement a form using Soil Form in my Android project, I noticed that there are many fields that need to be passed to the Composable in the Control.

```kt
    val control = rememberFieldRuleControl(
        name = "email",
        select = { this },
        update = { it },
    ) {
        notBlank { "not blank" }
    }

    Controller(control) { field ->
        TextField(
            value = field.value,
            onValueChange = field.onChange,
            placeholder = { Text(field.name) },
            modifier = Modifier
                .fillMaxWidth()
                .onFocusChanged(field),
            enabled = field.isEnabled,
            isError = field.hasError,
            singleLine = true,
            supportingText = {
                if (field.hasError) {
                    Text(text = field.errors.first(), color = MaterialTheme.colorScheme.error)
                }
            },
            keyboardOptions = KeyboardOptions(
                keyboardType = KeyboardType.Text,
                imeAction = ImeAction.Next,
            )
        )
    }
```

While we recognize that this is a great design from a flexibility standpoint, we think this may be too much work from a convenience standpoint.

## Suggestion.

We think that the burden on implementors can be reduced by providing helpers for detailed implementation of the commonly used `material` / `material3` libraries as follows.

```kt
// usage
val control = rememberFieldRuleControl(
  name = "email",
  select = { keyword },
  update = { copy(keyword = it) },
) {
  notBlank { "not blank" }
}

TextField(control)


// definition
fun TextField(
  control : FieldControl<String>,
  modifier: Modifier = Modifier,
  placeholder: @Composable (() -> Unit)? = { Text(field.name) }
  // ...
) {

    Controller(control) { field ->
        TextField(
            value = field.value,
            onValueChange = field.onChange,
            placeholder = placeholder,
            modifier = Modifier
                .onFocusChanged(field),
            // ...
        )
    }

}
```

I would like to discuss the pros and cons of this proposal first.

**Comment by TBSten** [comment]
If I were to proceed with the implementation, I would want to make sure the following are in place beforehand

## List of supported components

I would like to discuss whether there are over or under components in the subject.

- material
  - Buttons
  - Checkboxes
  - Radio buttons
  - Sliders
  - Switches
  - TextFields
- material3
  - Buttons
  - Checkbox
  - Chips
  - Date Picker
  - Radio button
  - Sliders
  - Switch
  - Text fields
  - Time Picker

- refs: 
  - [material components list](https://developer.android.com/reference/kotlin/androidx/compose/material/package-summary#components) 
  - [material3 components list](https://developer.android.com/reference/kotlin/androidx/compose/material3/package-summary#components) 


## Distribution method

I think it would be a good idea to release this helper as an independent library and allow users to opt-in to install it.

```
implementation("com.soil-kt.soil:form:$soilVersion")
implementation("com.soil-kt.soil:form-material:$soilVersion")

// or

implementation("com.soil-kt.soil:form:$soilVersion")
implementation("com.soil-kt.soil:form-material3:$soilVersion")
```

This is because some users may not want to rely on material design.

**Comment by ogaclejapan** [comment]
Thank you for your proposal, @TBSten 

I actually toyed with this idea before our Alpha release but ended up shelving it for now. 
I think we should hold off until we get a better sense of where Compose Multiplatform (CMP) is heading.

### The Future of the Compose UI Library is Still Unstable

Right now, both Android's Material (M2) and Material3 (M3) UI libraries are still being used just as widely, with no cross-compatibility in form-related components. For example, supportingText is only a thing in M3, and other components like ExposedDropDownMenuBox have different modifiers. It’s a hefty lift to keep pace with both libraries’ updates and new features 😅 

Also, looking at CMP, there are some community-driven efforts to create adaptive UI libraries for iOS. 
We might need to integrate similar things down the line.

e.g., https://github.com/alexzhirkevich/compose-cupertino

And with Wasm support picking up, we might start seeing some cool platform-agnostic components popping up.

**Comment by ogaclejapan** [comment]
In actual android app development, it is common to customize Material UI Form-related components to fit the project-specific design.

- Defining guidelines for common UI components
- Maintaining components in line with the design system

I recommend to build a layer of components for common use within the project, including fine adjustments.
The Form component in the playground module serves as an example of such usage. 

https://github.com/soil-kt/soil/tree/main/internal/playground/src/commonMain/kotlin/soil/playground/form/compose

It's very difficult to provide a UI that fits every project 🎨

**Comment by TBSten** [comment]
Thanks for the reply!

I think you are right that the compose UI is unstable. Its burdensome and we will have to wait a while to do it.

I wonder if one of the options for building UI in development with Compose could be a "UI library" like MUI or Chakra UI in React. I am thinking about it. This would be very beneficial for small app developers. Of course, this is not suitable for large scale app development.
The intent of my proposal is to add integration to Soil without lacking the extensibility of the existing Material Component. It is the library user who chooses whether or not to implement it, based on the scale of his or her application.

In summary, it seems to me that this issue should be Closed at once 🙃.

**Comment by ogaclejapan** [comment]
> I am thinking about it. 

Cool! I'm looking forward to seeing it completed your project 😄 

I believe the demand for headless components for Compose (like Ark UI or Radix UI) will increase in the future.
Currently, I'm a bit disappointed that there are no headless components available at the base layers of M2 and M3.
